models/issue.go

```diff
@@ -1498,3 +1507,27 @@ func updateIssue(e Engine, issue *Issue) error {
 func UpdateIssue(issue *Issue) error {
 	return updateIssue(x, issue)
 }
+
+// UpdateIssueDeadline updates an issue deadline and adds comments. Setting a deadline to 0 means deleting it.
+func UpdateIssueDeadline(issue *Issue, deadlineUnix util.TimeStamp, doer *User) (err error) {
+
+	// if the deadline hasn't changed do nothing
+	if issue.DeadlineUnix == deadlineUnix {
+		return nil
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+
+	// Update the deadline
+	if _, err = sess.ID(issue.ID).Cols("deadline_unix").Update(&Issue{DeadlineUnix: deadlineUnix}); err != nil {
+		return err
+	}
+
+	// Make the comment
+	if _, err = createDeadlineComment(sess, doer, issue, deadlineUnix); err != nil {
+		return fmt.Errorf("createRemovedDueDateComment: %v", err)
+	}
+
+	return sess.Commit()
```

Comment:
If you want transaction, you have to call `Begin()` first (from xorm doc):
```
if err := session.Begin(); err != nil {
    // if returned then will rollback automatically
    return err
}
```

If not, this is not needed and you can just `return nil`.

Reply:
Sure, I'll add that.